File: src/lib/providers/catalog.ts
export const PROVIDER_APIS = ["openai", "anthropic", "google"] as const;

export type ProviderApi = (typeof PROVIDER_APIS)[number];

export type VendorId =
  | "openai"
  | "anthropic"
  | "google"
  | "deepseek"
  | "zhipu"
  | "minimax"
  | "stepfun"
  | "moonshot"
  | "siliconflow"
  | "custom";

export interface VendorEndpoint {
  baseUrl: string;
  modelsPath?: string;
}

export interface VendorMeta {
  id: VendorId;
  labelKey: string;
  homepage?: string;
  apiKeyUrl?: string;
  endpoints: Partial<Record<ProviderApi, VendorEndpoint>>;
}

export const VENDOR_CATALOG: Readonly<Record<VendorId, VendorMeta>> = {
  openai: {
    id: "openai",
    labelKey: "providers.vendor.openai",
    homepage: "https://openai.com",
    apiKeyUrl: "https://platform.openai.com/api-keys",
    endpoints: { openai: { baseUrl: "https://api.openai.com/v1", modelsPath: "/models" } },
  },
  anthropic: {
    id: "anthropic",
    labelKey: "providers.vendor.anthropic",
    homepage: "https://anthropic.com",
    apiKeyUrl: "https://console.anthropic.com/settings/keys",
    endpoints: { anthropic: { baseUrl: "https://api.anthropic.com", modelsPath: "/v1/models" } },
  },
  google: {
    id: "google",
    labelKey: "providers.vendor.google",
    homepage: "https://ai.google.dev",
    apiKeyUrl: "https://aistudio.google.com/apikey",
    endpoints: { google: { baseUrl: "https://generativelanguage.googleapis.com", modelsPath: "/v1beta/models" } },
  },
  deepseek: {
    id: "deepseek",
    labelKey: "providers.vendor.deepseek",
    homepage: "https://deepseek.com",
    apiKeyUrl: "https://platform.deepseek.com/api_keys",
    endpoints: {
      openai: { baseUrl: "https://api.deepseek.com/v1", modelsPath: "/models" },
      anthropic: { baseUrl: "https://api.deepseek.com/anthropic" },
    },
  },
  zhipu: {
    id: "zhipu",
    labelKey: "providers.vendor.zhipu",
    homepage: "https://bigmodel.cn",
    apiKeyUrl: "https://bigmodel.cn/usercenter/proj-mgmt/apikeys",
    endpoints: {
      openai: { baseUrl: "https://open.bigmodel.cn/api/paas/v4", modelsPath: "/models" },
      anthropic: { baseUrl: "https://open.bigmodel.cn/api/anthropic" },
    },
  },
  minimax: {
    id: "minimax",
    labelKey: "providers.vendor.minimax",
    homepage: "https://minimaxi.com",
    apiKeyUrl: "https://platform.minimaxi.com/user-center/basic-information/interface-key",
    endpoints: { openai: { baseUrl: "https://api.minimaxi.com/v1", modelsPath: "/models" } },
  },
  stepfun: {
    id: "stepfun",
    labelKey: "providers.vendor.stepfun",
    homepage: "https://stepfun.com",
    apiKeyUrl: "https://platform.stepfun.com/interface-key",
    endpoints: { openai: { baseUrl: "https://api.stepfun.com/v1", modelsPath: "/models" } },
  },
  moonshot: {
    id: "moonshot",
    labelKey: "providers.vendor.moonshot",
    homepage: "https://moonshot.cn",
    apiKeyUrl: "https://platform.moonshot.cn/console/api-keys",
    endpoints: {
      openai: { baseUrl: "https://api.moonshot.cn/v1", modelsPath: "/models" },
      anthropic: { baseUrl: "https://api.moonshot.cn/anthropic" },
    },
  },
  siliconflow: {
    id: "siliconflow",
    labelKey: "providers.vendor.siliconflow",
    homepage: "https://siliconflow.cn",
    apiKeyUrl: "https://cloud.siliconflow.cn/account/ak",
    endpoints: { openai: { baseUrl: "https://api.siliconflow.cn/v1", modelsPath: "/models" } },
  },
  custom: {
    id: "custom",
    labelKey: "providers.vendor.custom",
    endpoints: {
      openai: { baseUrl: "" },
      anthropic: { baseUrl: "" },
      google: { baseUrl: "" },
    },
  },
};

export function supportedVendorIds(): VendorId[] {
  return Object.keys(VENDOR_CATALOG) as VendorId[];
}

export function getEndpoint(
  api: ProviderApi,
  vendor: VendorId,
): VendorEndpoint | undefined {
  return VENDOR_CATALOG[vendor].endpoints[api];
}

File: src-tauri/src/services/llm/catalog.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProviderApi {
    OpenAi,
    Anthropic,
    Google,
}

impl ProviderApi {
    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "openai" => Some(Self::OpenAi),
            "anthropic" => Some(Self::Anthropic),
            "google" => Some(Self::Google),
            _ => None,
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::OpenAi => "openai",
            Self::Anthropic => "anthropic",
            Self::Google => "google",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct VendorEndpoint {
    pub base_url: &'static str,
    pub models_path: Option<&'static str>,
}

pub fn supported_vendors() -> &'static [&'static str] {
    &[
        "openai",
        "anthropic",
        "google",
        "deepseek",
        "zhipu",
        "minimax",
        "stepfun",
        "moonshot",
        "siliconflow",
        "custom",
    ]
}

pub fn endpoint(api: ProviderApi, vendor: &str) -> Option<VendorEndpoint> {
    match (api, vendor) {
        (ProviderApi::OpenAi, "openai") => Some(openai("https://api.openai.com/v1")),
        (ProviderApi::Anthropic, "anthropic") => Some(anthropic("https://api.anthropic.com")),
        (ProviderApi::Google, "google") => {
            Some(google("https://generativelanguage.googleapis.com"))
        }
        (ProviderApi::OpenAi, "deepseek") => Some(openai("https://api.deepseek.com/v1")),
        (ProviderApi::Anthropic, "deepseek") => Some(no_list("https://api.deepseek.com/anthropic")),
        (ProviderApi::OpenAi, "zhipu") => Some(openai("https://open.bigmodel.cn/api/paas/v4")),
        (ProviderApi::Anthropic, "zhipu") => {
            Some(no_list("https://open.bigmodel.cn/api/anthropic"))
        }
        (ProviderApi::OpenAi, "minimax") => Some(openai("https://api.minimaxi.com/v1")),
        (ProviderApi::OpenAi, "stepfun") => Some(openai("https://api.stepfun.com/v1")),
        (ProviderApi::OpenAi, "moonshot") => Some(openai("https://api.moonshot.cn/v1")),
        (ProviderApi::Anthropic, "moonshot") => Some(no_list("https://api.moonshot.cn/anthropic")),
        (ProviderApi::OpenAi, "siliconflow") => Some(openai("https://api.siliconflow.cn/v1")),
        (_, "custom") => Some(no_list("")),
        _ => None,
    }
}

pub fn models_url(
    api: ProviderApi,
    vendor: &str,
    base_url_override: Option<&str>,
) -> Option<String> {
    let endpoint = endpoint(api, vendor)?;
    let models_path = endpoint.models_path?;
    let base_url = base_url_override
        .filter(|url| !url.trim().is_empty())
        .unwrap_or(endpoint.base_url);
    Some(join_url(base_url, models_path))
}

pub fn chat_url(api: ProviderApi, vendor: &str, base_url_override: Option<&str>) -> Option<String> {
    let endpoint = endpoint(api, vendor)?;
    let base_url = base_url_override
        .filter(|url| !url.trim().is_empty())
        .unwrap_or(endpoint.base_url);
    let path = match api {
        ProviderApi::OpenAi => "/chat/completions",
        ProviderApi::Anthropic => "/v1/messages",
        ProviderApi::Google => return None,
    };
    Some(join_url(base_url, path))
}

pub fn google_generate_url(base_url: Option<&str>, model_id: &str) -> String {
    let base = base_url
        .filter(|url| !url.trim().is_empty())
        .unwrap_or("https://generativelanguage.googleapis.com");
    let model = model_id.trim_start_matches("models/");
    format!(
        "{}/v1beta/models/{}:generateContent",
        base.trim_end_matches('/'),
        model
    )
}

fn openai(base_url: &'static str) -> VendorEndpoint {
    VendorEndpoint {
        base_url,
        models_path: Some("/models"),
    }
}

fn anthropic(base_url: &'static str) -> VendorEndpoint {
    VendorEndpoint {
        base_url,
        models_path: Some("/v1/models"),
    }
}

fn google(base_url: &'static str) -> VendorEndpoint {
    VendorEndpoint {
        base_url,
        models_path: Some("/v1beta/models"),
    }
}

fn no_list(base_url: &'static str) -> VendorEndpoint {
    VendorEndpoint {
        base_url,
        models_path: None,
    }
}

fn join_url(base_url: &str, path: &str) -> String {
    format!(
        "{}/{}",
        base_url.trim_end_matches('/'),
        path.trim_start_matches('/')
    )
}
